extensions/amp-analytics/0.1/visibility-manager.js

```diff
@@ -230,11 +231,69 @@ export class VisibilityManager {
    */
   listenElement(
       element, spec, readyPromise, createReportPromiseFunc, callback) {
-    const model = new VisibilityModel(
-        spec,
-        this.getElementVisibility.bind(this, element));
-    return this.listen_(
-        model, spec, readyPromise, createReportPromiseFunc, callback, element);
+    const calcVisibility = this.getElementVisibility.bind(this, element);
+    return this.createModelAndListen_(calcVisibility, spec, readyPromise,
+        createReportPromiseFunc, callback, element);
+  }
+
+  /**
+   * Create visibilityModel and listen to visible events.
+   * @param {function():number} calcVisibility
+   * @param {!Object<string, *>} spec
+   * @param {?Promise} readyPromise
+   * @param {?function():!Promise} createReportPromiseFunc
+   * @param {function(!Object<string, *)} callback
+   * @param {!Element=} opt_element
+   * @return {!UnlistenDef}
+   */
+  createModelAndListen_(calcVisibility, spec,
+      readyPromise, createReportPromiseFunc, callback, opt_element) {
+    if (spec['visiblePercentageThresholds'] &&
+        spec['visiblePercentageMin'] == undefined &&
+        spec['visiblePercentageMax'] == undefined) {
+      const unlisteners = [];
+      //const ranges = tryParseJson(spec['visiblePercentageThresholds']);
+      const ranges = spec['visiblePercentageThresholds'];
+      if (!ranges || !isArray(ranges)) {
+        user().error(TAG, 'invalid visiblePercentageThresholds');
+      }
+      console.log('ranges is ', ranges);
+      for (let i = 0; i < ranges.length; i++) {
+        const percents = ranges[i];
+        console.log('percents is ', percents);
+        if (!isArray(percents) || percents.length != 2) {
+          user().error(TAG,
+              'visiblePercentageThresholds entry length is not 2');
+          continue;
+        }
+        if (!isFiniteNumber(percents[0]) || !isFiniteNumber(percents[1])) {
+          // not valid number
+          user().error(TAG,
+              'visiblePercentageThresholds entry is not valid number');
+          continue;
+        }
+        const min = Number(percents[0]);
+        const max = Number(percents[1]);
+        if (min < 0 || max > 100 || min >= max) {
```

Comment:
Why can't min == max?

Reply:
IntersectionObserver require the visiblePercentage to be greater than min, and less or equal to max. With `min==max`, we would not be able to fire any visible event.

Reply:
Ah, I see. Thank you for clarifying!